index/segment/types.go

```diff
@@ -136,3 +99,60 @@ type Filter struct {
 	// Note: RE2 only, no PCRE (i.e. no backreferences).
 	Regexp bool
 }
+
+// ImmutablePostingsList is a memory efficient collection of docIDs. It only permits
+// for read-only operations.
+type ImmutablePostingsList interface {
+	// Contains returns whether the specified id is contained in the set.
+	Contains(i DocID) bool
+
+	// IsEmpty returns whether the set is empty.
+	IsEmpty() bool
+
+	// Size returns the numbers of ids in the set.
+	Size() uint64
+
+	// Iter returns an iterator over the known values.
+	Iter() PostingsIter
+
+	// Clone returns a copy of the set.
+	Clone() PostingsList
+}
+
+// PostingsList is a memory efficient collection of docIDs.
+type PostingsList interface {
```

Comment:
I think all the mutable operations here should probably return `error`.

Reply:
Sure thing.

Reply:
done